### PR TITLE
HTTPRoute, TCPRoute: Add gateway status information

### PR DESCRIPTION
**Pull request by Miciah**
* `apis/v1alpha1/httproute_types.go` (`HTTPRouteStatus`): Change the slice of `GatewayObjectReference` to a slice of `RouteGatewayStatus`.
(`RouteGatewayStatus`): New type.  Define a reference to a `Gateway` object and a list of status conditions for the route with respect to that `Gateway`.
(`RouteConditionType`): New type for route status condition type.
(`ConditionRouteAdmitted`): New route status condition type.  A route status condition with this type indicates whether the route has been admitted or rejected by a `Gateway`.
(`RouteCondition`): New type for route status conditions.
* `apis/v1alpha1/tcproute_types.go` (`TCPRouteStatus`): Add a `Gateways` field with type slice of `RouteGatewayStatus`.
* `apis/v1alpha1/generated.pb.go`:
* `apis/v1alpha1/generated.proto`:
* `apis/v1alpha1/zz_generated.deepcopy.go`:
* `config/crd/bases/networking.x-k8s.io_httproutes.yaml`:
* `config/crd/bases/networking.x-k8s.io_tcproutes.yaml`:
* `docs-src/spec.md`:
* `docs/spec/index.html`: Regenerate.

### CODE REVIEW

File: api/v1alpha1/httproute_types.go [hunk 1]
```diff
@@ -192,7 +192,52 @@ type HTTPRouteAction struct {
...
+	// this list.
+	Gateways []RouteGatewayStatus `json:"gateways"`
+}
+
+// RouteGatewayStatus describes the status of a route with respect to an
+// associated Gateway.
+type RouteGatewayStatus struct {
+	// Gateway is a reference to a Gateway object that is associated with
+	// the route.
+	Gateway core.ObjectReference `json:"gateway"`
```
Comment (Miciah) [inline api/v1alpha1/httproute_types.go]: This PR has `core.ObjectReference` for consistency with master; if #73 is accepted and this PR is not rejected, I will update this field's type.

File: api/v1alpha1/httproute_types.go [hunk 2]
```diff
@@ -192,7 +192,52 @@ type HTTPRouteAction struct {
 
 // HTTPRouteStatus defines the observed state of HTTPRoute
 type HTTPRouteStatus struct {
-	Gateways []core.ObjectReference `json:"gateways"`
+	// Gateways is a list of the Gateways that are associated with the
+	// HTTPRoute, and the status of the route with respect to each of these
+	// Gateways.  Each associated Gateway should have exactly one entry in
+	// this list.
+	Gateways []RouteGatewayStatus `json:"gateways"`
```
Comment (Miciah) [inline api/v1alpha1/httproute_types.go]: Open question:  What happens if a gateway controller adds a status condition to a route, and the controller is subsequently deleted?  What is responsible for clearing out old status conditions?

**Comment by Miciah** [comment]
Rebased.

**Comment by Miciah** [comment]
Rebased.

**Comment by robscott** [comment]
Sorry we missed this earlier @Miciah! This still looks relevant, will make sure we cover it today.

**Comment by Miciah** [comment]
Rebased.

File: apis/v1alpha1/httproute_types.go [hunk 3]
```diff
@@ -327,6 +332,47 @@ type GatewayObjectReference struct {
...
+// associated Gateway.
+type RouteGatewayStatus struct {
+	// Gateway is a reference to a Gateway object that is associated with
+	// the route.
+	Gateway GatewayObjectReference `json:"gateways" protobuf:"bytes,4,rep,name=gateways"`
+	// Conditions describes the status of the route with respect to the
+	// Gateway.  For example, the "Admitted" condition indicates whether the
+	// route has been admitted or rejected by the Gateway, and why.  Note
+	// that the route's availability is also subject to the Gateway's own
+	// status conditions and listener status, which see.
```
Comment (robscott) [inline apis/v1alpha1/httproute_types.go:345]: I think this may have gotten cut off early?

File: apis/v1alpha1/httproute_types.go [hunk 4]
```diff
@@ -312,7 +313,11 @@ type RouteHostExtensionObjectReference = ConfigMapsDefaultLocalObjectReference
 
 // HTTPRouteStatus defines the observed state of HTTPRoute.
 type HTTPRouteStatus struct {
-	GatewayRefs []GatewayObjectReference `json:"gatewayRefs" protobuf:"bytes,1,rep,name=gatewayRefs"`
+	// Gateways is a list of the Gateways that are associated with the
+	// HTTPRoute, and the status of the route with respect to each of these
+	// Gateways.  Each associated Gateway should have exactly one entry in
+	// this list.
```
Comment (robscott) [inline apis/v1alpha1/httproute_types.go:319]: We should add some kind of clarification as to what controllers should or must do here, assuming controllers are responsible for setting this status.

File: apis/v1alpha1/httproute_types.go (hunk 3 above)
Comment (Miciah) [reply apis/v1alpha1/httproute_types.go:345]: I'm not sure what you mean.  Are you referring to "which see"?  It's an admittedly unusual construct (similar to ["q.v."](https://en.wiktionary.org/wiki/q.v.)).  Maybe best just to delete it.

File: apis/v1alpha1/httproute_types.go (hunk 4 above)
Comment (Miciah) [reply apis/v1alpha1/httproute_types.go:319]: Basically the controller should set the status condition to either `True` or `False` when it sees the route, and possibly update that status condition when the route is modified.  I can add a statement to that effect.  I did have a question about cleaning up old status entries if a controller that has added a status condition is subsequently deleted (https://github.com/kubernetes-sigs/service-apis/pull/86#discussion_r379168521); this is still an open question that I'm not sure how best to address.

File: api/v1alpha1/httproute_types.go (hunk 2 above)
Comment (hbagdi) [reply api/v1alpha1/httproute_types.go]: Some of the existing ingress controllers do have a tunable setting [https://kubernetes.github.io/ingress-nginx/user-guide/cli-arguments/](https://kubernetes.github.io/ingress-nginx/user-guide/cli-arguments/) to control this behavior.
We should either suggest all implementers to take the same route or leave this out for v1alpha1 (by adding an issue to track it) and come back to it later.

**Comment by hbagdi** [comment]
This looks good to me overall.

> I'm curious if this would be difficult for controllers to keep up-to-date though.

I don't think so. While I've not worked out the details, this is similar to Ingress v1beta1/v1 status field which the controllers are responsible for keeping up to date and most implementations do it fine. We will need more condition types in future as we gain more experience in implementing and using this API.

File: apis/v1alpha1/httproute_types.go [hunk 5]
```diff
@@ -327,6 +332,47 @@ type GatewayObjectReference struct {
 	Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
 }
 
+// RouteGatewayStatus describes the status of a route with respect to an
+// associated Gateway.
+type RouteGatewayStatus struct {
+	// Gateway is a reference to a Gateway object that is associated with
+	// the route.
+	Gateway GatewayObjectReference `json:"gateways" protobuf:"bytes,4,rep,name=gateways"`
```
Comment (jpeach) [inline apis/v1alpha1/httproute_types.go:340]: I think this field should be `GatewayRef` to follow the API conventions recommendations. Update tag names to match,

File: apis/v1alpha1/httproute_types.go [hunk 6]
```diff
@@ -327,6 +332,47 @@ type GatewayObjectReference struct {
...
+	Conditions []RouteCondition `json:"conditions,omitempty" protobuf:"bytes,3,rep,name=conditions"`
+}
+
+// RouteConditionType is a type of condition for a route.
+type RouteConditionType string
+
+const (
+	// ConditionRouteAdmitted indicates whether the route has been admitted
+	// or rejected by a Gateway, and why.
+	ConditionRouteAdmitted RouteConditionType = "Admitted"
```
Comment (jpeach) [inline apis/v1alpha1/httproute_types.go:355]: I like the idea of this, but wanted to point out that this is a positive polarity condition, whereas all our other conditions are negative polarity. Not necessarily a bad thing, but we may wish to revisit after feedback.

Comment (hbagdi) [reply apis/v1alpha1/httproute_types.go:357]: Miciah, did you consider using the "Ready" condition for this?

Comment (bowei) [reply apis/v1alpha1/httproute_types.go:357]: At this point, I think we should just do what is natural, given the conversation in upstream. It is alpha and these can be fixed up in a straightforward manner.

Comment (Miciah) [reply apis/v1alpha1/httproute_types.go:357]: "Ready" could be misleading:  The controller may have admitted the route but have some unrelated configuration error that prevents the route from being accessible, or the controller may need additional time to make the route accessible (for example, restarting the proxy to reload the configuration).

**Pull request merged**